Add support for auth tokens to gateway and callback connections.

vanzin opened the pull request:
This change adds authentication to both the gateway and callback
servers, in the form of a shared authentication token.

The generation and distribution of tokens is left for applications
to do. The library just provides the means for enforcing the tokens.
The only exception is when using the gateway server command line,
in which case it will generate a token when requested.

By default, like previously, there is no authentication.

The auth protocol is very, very simple: when auth is enabled, the
servers expect the first line of content from the client connection
to be the auth token. If the token doesn't match, the server will
shut down the connection. Note this means that the token is sent
in plain text over the connection, meaning that for security it
requires the use of a local socket, or SSL for remote sockets.

Fixes #268

vanzin commented:
@bartdag gentle ping.

vanzin commented:
Not sure how to retrigger the test without making some dummy change...

bartdag commented on py4j-java/src/main/java/py4j/CallbackConnection.java:
```diff
@@ -207,6 +227,18 @@ public void start() throws IOException {
 		reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), Charset.forName("UTF-8")));
 		writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream(), Charset.forName("UTF-8")));
+
+		if (authToken != null) {
```
maybe extract in a private method?

bartdag commented on py4j-java/src/main/java/py4j/GatewayConnection.java:
```diff
@@ -203,6 +210,19 @@ public void run() {
 		boolean reset = false;
 		Throwable error = null;
 		try {
+			if (authToken != null) {
```
same comment as in clientconnection

bartdag commented on py4j-java/src/main/java/py4j/Protocol.java:
```diff
@@ -331,6 +331,10 @@ public final static String getOutputErrorCommand() {
 	}
 
+	public final static String getOutputSuccessCommand() {
+		return "" + RETURN_MESSAGE + SUCCESS + END_OUTPUT;
```
we usually use Protocol.getOutputVoidCommand();

bartdag commented on py4j-java/src/main/java/py4j/StringUtil.java:
```diff
@@ -33,16 +33,20 @@
 
 	public static String escape(String original) {
-		return original.replace("\\", "\\\\").replace("\r", "\\r").replace("\n", "\\n");
+		if (original != null) {
```
why is it necessary to add a check here considering that unescape has no such check?

vanzin replied:
I only use `escape` for the auth token, since it's a string and someone might add a new line there and break things. I don't need unescape for it.

This avoids adding another method to do a null-safe escape.

bartdag commented on py4j-python/src/py4j/protocol.py:
```diff
@@ -196,11 +199,14 @@ def unescape_new_line(escaped):
-            ("\r".join(p.split(ESCAPE_CHAR + "r")))
-            .split(ESCAPE_CHAR + "n"))
-        for p in escaped.split(ESCAPE_CHAR + ESCAPE_CHAR))
+    if escaped:
```
same comment as in StringUtil

vanzin replied:
Same reason.

vanzin commented:
@bartdag I think I addressed all your comments.

I left the "authentication as a command" changes in a separate commit since I'm not so sure that's a useful change. In my view, it makes the code more complicated and abuses the command abstraction, and doesn't really seem to make things any better overall. But your call.

bartdag commented:
Thanks for your latest commits, they do address all my concerns. I will probably have a shot at refactoring the way the auth command is initialized/called, but it makes the communication easier to debug (especially with the error message you provided). I was expecting the contribution over the weekend or early next week and had planned to work on the release during the April 28th weekend. I'll see what I can do during the weekend to at least make sure that everything is ready for a release.

Thanks a lot for your contribution and for taking the time to make your changes backward compatible: that means a lot to me and to Py4J users.

bartdag commented:
Just a heads up about the integration of this patch: at this point, I'm happy with the patch and the small refactoring I did. While testing though, I noticed that errors related to a missing token or a wrong token are not 100% obvious if logging is turned off so I'll need to spend a bit more time tomorrow trying to make it more obvious. I get many many support requests for simple errors such as a wrong ip address or a wrong port, so I am confident that wrong tokens will significantly increase my support load if it's not more obvious.

bartdag commented:
I will also fix the copyright year. There is a license template file that needs to be changed and spotlessApply will fix all copyright headers.
